synapse/logging/context.py

```diff
@@ -825,6 +825,8 @@ def defer_to_threadpool(reactor, threadpool, f, *args, **kwargs):
     logcontext = current_context()
 
     def g():
+        # It is not expected that this gets called from the sentinel context.
+        assert isinstance(logcontext, LoggingContext)
```

Comment:
we should probably run this on the calling thread rather than in the threadpool?

Reply:
Annoyingly mypy doesn't seem to like that. I wonder if this is another place where we should convert the sentinel into `None` and warn though?

Reply:
I reworked this code and this is now done on the calling thread.